common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -118,76 +119,433 @@ public Collation(
     }
 
     /**
-     * Constructor with comparators that are inherited from the given collator.
+     * Collation id is defined as 32-bit integer.
+     * We specify binary layouts for different classes of collations.
+     * Classes of collations are differentiated by most significant 3 bits (bit 31, 30 and 29),
+     * bit 31 being most significant and bit 0 being least significant.
+     * ---
+     * INDETERMINATE collation id binary layout:
+     * bit 31-0: 1
+     * INDETERMINATE collation id is equal to -1
+     * ---
+     * user-defined collation id binary layout:
+     * bit 31:   0
+     * bit 30:   1
+     * bit 29-0: undefined, reserved for future use
+     * ---
+     * UTF8_BINARY collation id binary layout:
+     * bit 31-22: zeroes
+     * bit 21-18: zeroes, reserved for space trimming
+     * bit 17-16: zeroes, reserved for version
+     * bit 15-3:  zeroes
+     * bit 2:     0, reserved for accent sensitivity
+     * bit 1:     0, reserved for uppercase and case-insensitive
+     * bit 0:     0 = case-sensitive, 1 = lowercase
+     * ---
+     * ICU collation id binary layout:
+     * bit 31-30: zeroes
+     * bit 29:    1
+     * bit 28-24: zeroes
+     * bit 23-22: zeroes, reserved for version
```

Comment:
Is there a good reason why version appears at different positions for the UTF8_BINARY family and ICU collations?

Reply:
I think it would be helpful (as part of a follow up PR) to provide a bit-centric view (right now we have a collation centric view.
Bit centric view:
```
31: 1 for INDETERMINATE (requires all other bits to be 1 as well), 0 for all other collations.
30: 0 for predefined, 1 for user-defined
29-24: Reserved
23-22: Reserved for version (here we have a discrepancy between UTF8_BINARY and ICU family).
21-18: Reserved for space trimming
17-0: Depend on collation family.
```

Reply:
Thanks for catching this. I refactored the comment to unify bits used for versioning and space trimming in both `UTF8_BINARY` and ICU collations. Added bit-centric view as well.